src/utils/saml.ts

```diff
@@ -1118,395 +82,557 @@ export const getPreValidateResponse = (
     "Response"
   );
 
-  const hasStrictValidation = fromNullable(strictValidationOptions)
-    .chain(_ => getSamlIssuer(doc).mapNullable(issuer => _[issuer]))
-    .getOrElse(false);
-
-  fromEitherToTaskEither(
-    fromPredicate<Error, HTMLCollectionOf<Element>>(
-      _ => _.length < 2,
-      _ => new Error("SAML Response must have only one Response element")
-    )(responsesCollection)
-      .map(_ => _.item(0))
-      .chain(Response =>
-        fromOption(new Error("Missing Reponse element inside SAML Response"))(
-          fromNullable(Response)
-        )
+  const hasStrictValidation = pipe(
+    O.fromNullable(strictValidationOptions),
+    O.chain(_ =>
+      pipe(
+        getSamlIssuer(doc),
+        O.chainNullableK(issuer => _[issuer])
       )
-      .chain(Response =>
-        mainAttributeValidation(Response, samlConfig.acceptedClockSkewMs).map(
-          IssueInstant => ({
+    ),
+    O.getOrElse(() => false)
+  );
+
+  interface IBaseOutput {
+    InResponseTo: NonEmptyString;
+    Assertion: Element;
+    IssueInstant: Date;
+    Response: Element;
+    AssertionIssueInstant: Date;
+  }
+
+  interface ISamlCacheType {
+    RequestXML: string;
+    createdAt: Date;
+    idpIssuer: string;
+  }
+
+  type IStepTwo = IBaseOutput & {
```

Comment:
Don't use generic `StepX` types names, use instead semantic types names to clearly identify what these types are.

Reply:
fixed!